File: src/go/pkg/domain/tier/tier.go
package tier

import (
	"time"

	pb "github.com/fitglue/server/src/go/pkg/types/pb"
)

const (
	HobbyistTierSyncsPerMonth  = 25
	HobbyistTierMaxConnections = 2
)

// Effective tier is used for internal logic
type EffectiveTier string

const (
	TierHobbyist EffectiveTier = "hobbyist"
	TierAthlete  EffectiveTier = "athlete"
)

// GetEffectiveTier determines the user's effective tier based on admin status,
// trial period, and stored tier.
func GetEffectiveTier(user *pb.UserRecord) EffectiveTier {
	// Admin override always grants Athlete
	if user.IsAdmin {
		return TierAthlete
	}

	// Active trial grants Athlete
	if user.TrialEndsAt != nil && user.TrialEndsAt.AsTime().After(time.Now()) {
		return TierAthlete
	}

	// Fall back to stored tier (default: hobbyist)
	if user.Tier == pb.UserTier_USER_TIER_ATHLETE {
		return TierAthlete
	}

	return TierHobbyist
}

// CanSync checks if user can perform a sync within their tier limits.
func CanSync(user *pb.UserRecord) (allowed bool, reason string) {
	tier := GetEffectiveTier(user)

	if tier == TierAthlete {
		return true, ""
	}

	// Check monthly limit for hobbyist tier
	if user.SyncCountThisMonth >= HobbyistTierSyncsPerMonth {
		return false, "Hobbyist tier limit reached (25/month). Upgrade to Athlete for unlimited syncs."
	}

	return true, ""
}

// CanAddConnection checks if user can add a new connection within their tier limits.
func CanAddConnection(user *pb.UserRecord, currentCount int) (allowed bool, reason string) {
	tier := GetEffectiveTier(user)

	if tier == TierAthlete {
		return true, ""
	}

	if currentCount >= HobbyistTierMaxConnections {
		return false, "Hobbyist tier limited to 2 connections. Upgrade to Athlete for unlimited."
	}

	return true, ""
}

// ShouldResetSyncCount checks if the sync counter should be reset (monthly)
func ShouldResetSyncCount(user *pb.UserRecord) bool {
	if user.SyncCountResetAt == nil {
		return true
	}

	resetTime := user.SyncCountResetAt.AsTime()
	now := time.Now()

	// Reset if the reset date is in a different month
	return resetTime.Year() != now.Year() || resetTime.Month() != now.Month()
}

// GetTrialDaysRemaining returns the number of days left in trial, or -1 if not on trial
func GetTrialDaysRemaining(user *pb.UserRecord) int {
	if user.TrialEndsAt == nil {
		return -1
	}

	now := time.Now()
	trialEnd := user.TrialEndsAt.AsTime()

	if trialEnd.Before(now) || trialEnd.Equal(now) {
		return 0
	}

	return int(trialEnd.Sub(now).Hours()/24) + 1
}

File: src/typescript/shared/src/domain/tier.ts
import { UserRecord, UserTier } from '../types/pb/user';
import { UserStore } from '../storage/firestore/user-store';
import { HttpError } from '../errors';

export const TIER_HOBBYIST = 'hobbyist' as const;
export const TIER_ATHLETE = 'athlete' as const;

export type EffectiveTier = typeof TIER_HOBBYIST | typeof TIER_ATHLETE;

export const HOBBYIST_TIER_LIMITS = {
  SYNCS_PER_MONTH: 25,
  MAX_CONNECTIONS: 2,
} as const;

/**
 * Determine the effective tier for a user.
 * Priority: admin > active trial > stored tier
 */
export function getEffectiveTier(user: UserRecord): EffectiveTier {
  // Admin override always grants Athlete
  if (user.isAdmin) {
    return TIER_ATHLETE;
  }

  // Active trial grants Athlete
  if (user.trialEndsAt && new Date(user.trialEndsAt) > new Date()) {
    return TIER_ATHLETE;
  }

  // Fall back to stored tier (default: hobbyist)
  if (user.tier === UserTier.USER_TIER_ATHLETE) {
    return TIER_ATHLETE;
  }

  return TIER_HOBBYIST;
}

/**
 * Check if a user can perform a sync (Free tier: 25/month limit)
 */
export function canSync(user: UserRecord): { allowed: boolean; reason?: string } {
  const tier = getEffectiveTier(user);

  if (tier === TIER_ATHLETE) {
    return { allowed: true };
  }

  // Check monthly limit
  const count = user.syncCountThisMonth || 0;
  if (count >= HOBBYIST_TIER_LIMITS.SYNCS_PER_MONTH) {
    return {
      allowed: false,
      reason: `Hobbyist tier limit reached (${HOBBYIST_TIER_LIMITS.SYNCS_PER_MONTH}/month). Upgrade to Athlete for unlimited syncs.`,
    };
  }

  return { allowed: true };
}

/**
 * Check if a user can add a new connection (Free tier: 2 max)
 */
export function canAddConnection(user: UserRecord, currentConnectionCount: number): { allowed: boolean; reason?: string } {
  const tier = getEffectiveTier(user);

  if (tier === TIER_ATHLETE) {
    return { allowed: true };
  }

  if (currentConnectionCount >= HOBBYIST_TIER_LIMITS.MAX_CONNECTIONS) {
    return {
      allowed: false,
      reason: `Hobbyist tier limited to ${HOBBYIST_TIER_LIMITS.MAX_CONNECTIONS} connections. Upgrade to Athlete for unlimited.`,
    };
  }

  return { allowed: true };
}

/**
 * Calculate trial days remaining
 */
export function getTrialDaysRemaining(user: UserRecord): number | null {
  if (!user.trialEndsAt) return null;

  const now = new Date();
  const trialEnd = new Date(user.trialEndsAt);

  if (trialEnd <= now) return 0;

  const diffMs = trialEnd.getTime() - now.getTime();
  return Math.ceil(diffMs / (1000 * 60 * 60 * 24));
}

/**
 * Count active integrations for a user
 */
export function countActiveConnections(user: UserRecord): number {
  const integrations = user.integrations || {};
  let count = 0;
  if (integrations.hevy?.enabled) count++;
  if (integrations.strava?.enabled) count++;
  if (integrations.fitbit?.enabled) count++;
  // Add mock only if in dev mode? For now, exclude from user-facing counts
  return count;
}

/**
 * Fetch a user via UserStore (which uses the Firestore converter for correct
 * snake_case → camelCase mapping) and throw 403 if their effective tier is
 * not Athlete.
 */
export async function requireAthleteTier(
  userStore: UserStore,
  userId: string
): Promise<UserRecord> {
  const user = await userStore.get(userId);
  if (!user) {
    throw new HttpError(404, 'User not found');
  }
  if (getEffectiveTier(user) !== TIER_ATHLETE) {
    throw new HttpError(403, 'This functionality requires Athlete tier');
  }
  return user;
}

